chpl/chpl-service/src/main/java/gov/healthit/chpl/report/criteriamigrationreport/CriteriaMigrationDefinitionEntity.java

```diff
@@ -0,0 +1,78 @@
+package gov.healthit.chpl.report.criteriamigrationreport;
+
+import java.util.List;
+
+import javax.persistence.Basic;
+import javax.persistence.Column;
+import javax.persistence.Entity;
+import javax.persistence.FetchType;
+import javax.persistence.GeneratedValue;
+import javax.persistence.GenerationType;
+import javax.persistence.Id;
+import javax.persistence.JoinColumn;
+import javax.persistence.OneToMany;
+import javax.persistence.OneToOne;
+import javax.persistence.Table;
+
+import org.hibernate.Hibernate;
+
+import gov.healthit.chpl.certificationCriteria.CertificationCriterionEntity;
+import gov.healthit.chpl.entity.EntityAudit;
+import gov.healthit.chpl.entity.lastmodifieduserstrategy.LastModifiedUserStrategy;
+import gov.healthit.chpl.entity.lastmodifieduserstrategy.SystemUserStrategy;
+import lombok.AllArgsConstructor;
+import lombok.Getter;
+import lombok.NoArgsConstructor;
+import lombok.Setter;
+import lombok.ToString;
+import lombok.experimental.SuperBuilder;
+
+@Getter
+@Setter
+@ToString
+@SuperBuilder
+@AllArgsConstructor
+@NoArgsConstructor
+@Entity
+@Table(name = "criteria_migration_definition")
+public class CriteriaMigrationDefinitionEntity extends EntityAudit {
+    private static final long serialVersionUID = 5484388441051382895L;
+
+    @Override
+    public LastModifiedUserStrategy getLastModifiedUserStrategy() {
+        return new SystemUserStrategy();
+    }
+
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    @Basic(optional = false)
+    @Column(name = "id", nullable = false)
+    private Long id;
+
+    @Basic(optional = false)
+    @Column(name = "criteria_migration_report_id", nullable = false)
+    private Long criteriaMigrationReportId;
+
+    @OneToOne(optional = false, fetch = FetchType.LAZY)
+    @JoinColumn(name = "original_certification_criterion_id", insertable = false, updatable = false)
+    private CertificationCriterionEntity originalCriterion;
+
+    @OneToOne(optional = false, fetch = FetchType.LAZY)
+    @JoinColumn(name = "updated_certification_criterion_id", insertable = false, updatable = false)
+    private CertificationCriterionEntity updatedCriterion;
+
+    @OneToMany(fetch = FetchType.LAZY, mappedBy = "id")
+    @Column(name = "criteria_migration_definition_id", nullable = false)
+    private List<CriteriaMigrationCountEntity> criteriaMigrationCounts;
+
+    public CriteriaMigrationDefinition toDomain() {
+        return CriteriaMigrationDefinition.builder()
+                .id(this.id)
+                .originalCriterion(this.originalCriterion.toDomain())
+                .updatedCriterion(this.updatedCriterion.toDomain())
+                .criteriaMigrationCounts(!Hibernate.isInitialized(this.criteriaMigrationCounts) ? null : this.criteriaMigrationCounts.stream()
```

Comment:
I've never seen this Hibernate.isInitialized method used before. Is there something unique about this entity or the HQL that gets it that makes this necessary? Is it like checking if there's a session before it would lazily try to fetch data? I think we should write our queries/code so that never happens right? Just wanted to understand this.

Reply:
Yes - there are some cases where I just get the structure of the report: CriteriaMigrationReport -> CriteriaMigrationDefinitions.

Then, sometimes I need to the structure of the report and the associated data we have collected:  CriteriaMigrationReport -> CriteriaMigrationDefinitions -> CriteriaMigrationCounts 

So, there are instances where the "counts" were not fetched intentionally.  When I intentionally did not fetch the counts, I would get transaction errors since Hibernate would try to retrieve the counts on this particular line.  The `Hibernate.isInitalized` seemed to handle the problem.  

Might have gotten the answer from here:  https://stackoverflow.com/questions/1589144/is-there-a-helper-to-know-whether-a-property-has-been-loaded-by-hibernate